### PR TITLE
Add search A/B test

### PR DESCRIPTION
**Pull request by jondewijones**
### What

Add search A/B test

### Who can review

Anyone

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -66,7 +68,18 @@ func Get() (*Config, error) {
...
 		ZebedeeRequestMaximumTimeoutSeconds: 5 * time.Second,
 		ZebedeeRequestMaximumRetries:        0,
+		EnableSearchABTest:                  false,
+		SearchABTestPercentage:              10,
 	}
 
 	return cfg, envconfig.Process("", cfg)
 }
+
+// IsEnableSearchABTest checks wehther ab test is enabled and that percentage is a sensuible int value
```
Comment (nshumoogum) [inline config/config.go:78]: Typos:
```suggestion
// IsEnableSearchABTest checks whether ab test is enabled and that percentage is a sensible int value
```

File: handlers/abtest/search.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+	"github.com/ONSdigital/dp-cookies/cookies"
+	"github.com/ONSdigital/log.go/log"
+)
+
+//go:generate moq -out searchtest/handler.go -pkg test . Handler
+type Handler http.Handler
+
+func SearchHandler(newSearch, oldSearch http.Handler, percentage int, domain string) http.Handler {
+	rand.Seed(time.Now().UnixNano())
```
Comment (nshumoogum) [inline handlers/abtest/search.go:16]: I assume this line can be removed

File: handlers/abtest/search.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		// Retrieve AB Test Cookie
+		cookie, err := cookies.GetABTest(req)
+		if err != nil && err != cookies.ErrABTestCookieNotFound {
+			log.Event(req.Context(), "error getting a/b test cookie", log.WARN, log.Error(err))
+		}
+
+		// If AB Test cookie not set, create new AB Test cookie
+		if cookie.NewSearch == nil && cookie.OldSearch == nil {
```
Comment (nshumoogum) [inline handlers/abtest/search.go:26]: Both If statements on line 26 and 39 contain the same code - maybe try to collapse it into a separate function that can be called by both if statements.

File: handlers/abtest/search.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		oldSearch = setTime24HoursAhead(now)
+	}
+
+	return cookies.ABServices{
+		NewSearch: &newSearch,
+		OldSearch: &oldSearch,
+	}
+}
+
+func servABTest(newSearch, oldSearch http.Handler, w http.ResponseWriter, req *http.Request, cookie cookies.ABServices) {
```
Comment (nshumoogum) [inline handlers/abtest/search.go:76]: There is a possibility that the old or new search cookie might have expired between calculating if old search or new search has expired and setting a new time value, prior to calling this function. I think it is better to assume old search if this is the case, like so:

```suggestion
func servABTest(newSearch, oldSearch http.Handler, w http.ResponseWriter, req *http.Request, cookie cookies.ABServices) {
	if cookie.NewSearch.After(time.Now().UTC()) {
		newSearch.ServeHTTP(w, req)
	} else {
		oldSearch.ServeHTTP(w, req)
	}
}
```

Comment (nshumoogum) [reply handlers/abtest/search.go:76]: Or you could pass time.Now().UTC() value used in previous function to determine this. E.g. set `currentTime := time.Now().UTC()` in handler and reuse by passing it into functions.

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: config/config.go (hunk 1 above)
Comment (jondewijones) [reply config/config.go:78]: 👍

File: handlers/abtest/search.go (hunk 2 above)
Comment (jondewijones) [reply handlers/abtest/search.go:16]: I think it's needed for truly random numbers?

>Top-level functions, such as Float64 and Int, use a default shared Source that produces a deterministic sequence of values each time a program is run. Use the Seed function to initialize the default Source if different behavior is required for each run.

File: handlers/abtest/search.go (hunk 3 above)
Comment (jondewijones) [reply handlers/abtest/search.go:26]: 👍

File: handlers/abtest/search.go (hunk 4 above)
Comment (jondewijones) [reply handlers/abtest/search.go:76]: 👍

File: handlers/abtest/search.go (hunk 2 above)
Comment (nshumoogum) [reply handlers/abtest/search.go:16]: Ah okay, I wasn't sure how this was being used but rand is the receiver for Seed method.

Could I suggest moving this line to where we act on the receiver on line 64 of this file?

**Review by nshumoogum — APPROVED** [review]

**Pull request merged**